test/QueryHubTest.cpp

```diff
@@ -104,59 +104,46 @@ ASYNC_TEST(QueryHub, simulateLifecycleWithDifferentQueryIds) {
 
 // _____________________________________________________________________________
 
-ASYNC_TEST_N(QueryHub, testCorrectReschedulingForEmptyPointerOnSignalEnd, 3) {
-  QueryHub queryHub{ioContext};
-  QueryId queryId = QueryId::idFromString("abc");
-
-  auto distributor1 =
-      co_await queryHub.createOrAcquireDistributorForSending(queryId);
-
-  co_await net::dispatch(
-      net::bind_executor(queryHub.getStrand(), net::use_awaitable));
-  auto future =
-      net::co_spawn(ioContext, distributor1->signalEnd(), net::use_future);
-
-  // Wait until signalEnd() blocks, increase time if tests sporadically fail
-  std::this_thread::sleep_for(std::chrono::milliseconds(2));
-
-  auto distributor2 =
-      co_await queryHub.createOrAcquireDistributorForSending(queryId);
-  EXPECT_NE(distributor1, distributor2);
-
-  future.wait();
-}
-
-// _____________________________________________________________________________
+namespace ad_utility::websocket {
 
 ASYNC_TEST_N(QueryHub, testCorrectReschedulingForEmptyPointerOnDestruct, 2) {
   QueryHub queryHub{ioContext};
   QueryId queryId = QueryId::idFromString("abc");
+  std::atomic_flag flag{};
 
   auto distributor =
       co_await queryHub.createOrAcquireDistributorForReceiving(queryId);
   std::weak_ptr<const QueryToSocketDistributor> comparison = distributor;
 
   co_await net::dispatch(
-      net::bind_executor(queryHub.getStrand(), net::use_awaitable));
-  auto future = net::post(ioContext,
-                          std::packaged_task<void()>(
-                              [distributor = std::move(distributor)]() mutable {
-                                // Invoke destructor while the strand of
-                                // queryHub is still in use
-                                distributor.reset();
-                              }));
-
-  // Wait until destructor of distributor blocks, increase time if tests
-  // sporadically fail
-  std::this_thread::sleep_for(std::chrono::milliseconds(2));
+      net::bind_executor(queryHub.globalStrand_, net::use_awaitable));
+  auto future = net::post(
+      ioContext, std::packaged_task<void()>(
+                     [distributor = std::move(distributor), &flag]() mutable {
+                       // Invoke destructor while the strand of
+                       // queryHub is still in use
+                       flag.test_and_set();
+                       flag.notify_all();
+                       distributor.reset();
+                     }));
+
+  // Wait until destructor of distributor blocks, increase sleep duration if
+  // tests sporadically fail, because technically the flag needs to be
+  // triggered inside the destructor, not before it.
+  flag.wait(false);
+  std::this_thread::sleep_for(std::chrono::milliseconds{1});
+
   distributor =
-      co_await queryHub.createOrAcquireDistributorForReceiving(queryId);
+      co_await queryHub.createOrAcquireDistributorInternalUnsafe<false>(
+          queryId);
```

Comment:
I somewhat understand why this test cannot be simpler/using the public interfaces.
Is this a test the reliably deadlocks without the `poll_one` mechanism?
And couldn't this also be tested without the `unsafe` mechanisms, but by having this test loop run on a carefully chosen number of threads s.t. we have one thread too few without the `poll_one`.